api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+	TrustDomain string `json:"trustDomain,omitempty"`
+
+	// certificateDNSNames contains the additional DNS names to be added to the istiod certificate SAN.
+	// +listType=set
+	// +optional
+	CertificateDNSNames []string `json:"certificateDNSNames,omitempty"`
+
+	// certificateDuration is the istio-csr and the istiod certificates validity period.
+	// +kubebuilder:default:="1h"
+	// +optional
+	CertificateDuration *metav1.Duration `json:"certificateDuration,omitempty"`
+
+	// certificateRenewBefore is the ahead time to renew the istio-csr and the istiod certificates
+	// before expiry.
+	// +kubebuilder:default:="30m"
+	// +optional
+	CertificateRenewBefore *metav1.Duration `json:"certificateRenewBefore,omitempty"`
+
+	// privateKeySize is the istio-csr and the istiod certificate's key size. When the SignatureAlgorithm
+	// is RSA, must be >= 2048 and for ECDSA, can only be 256 or 384, corresponding to P-256 and P-384 respectively.
+	// +kubebuilder:default:=2048
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="privateKeySize is immutable once set"
+	// +optional
+	PrivateKeySize int `json:"privateKeySize,omitempty"`
+
+	// signatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="signatureAlgorithm is immutable once set"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration *metav1.Duration `json:"maxCertificateDuration,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for obtaining the certificates.
+type ServerConfig struct {
+	// port to serve istio-csr gRPC service.
+	// +kubebuilder:default:=443
+	// +kubebuilder:validation:XValidation:rule="oldSelf == 0 || self == oldSelf",message="port is immutable once set"
+	// +optional
+	Port int32 `json:"port,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +listType=atomic
+	// +kubebuilder:default:={"default"}
+	// +kubebuilder:validation:XValidation:rule="self.all(x, x in oldSelf) && oldSelf.all(x, x in self)",message="revisions is immutable once set"
+	// +kubebuilder:validation:MaxItems=10
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+
+	// namespace of the istio control-plane.
+	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="namespace is immutable once set"
```

Comment:
As the field is required, do we need to check `oldSelf == ''`?

Reply:
Yeah agree, its not required. I will update it.

Reply:
suggestion incorporated.